server/src/main/java/org/elasticsearch/indices/recovery/plan/ShardSnapshotsService.java

```diff
@@ -128,11 +129,30 @@ public void fetchLatestSnapshotsForShard(ShardId shardId, ActionListener<Optiona
                 .map(BlobStoreIndexShardSnapshot.FileInfo::metadata)
                 .collect(Collectors.toMap(StoreFileMetadata::name, Function.identity()));
 
-            InMemoryDirectory directory = new InMemoryDirectory(snapshotFiles);
+            // If the snapshot is taken using a Lucene version that this node cannot read
+            // (i.e. the snapshot was taken in a node with version > than this node version)
+            // reading the segment commit information could likely fail and we won't be able
+            // to recover from a snapshot. This should be a rare edge-case since for most cases
+            // the allocation deciders won't allow allocating replicas in nodes with older versions.
```

Comment:
I am not sure I follow this. The allocation deciders will not allow allocating replicas to an older node. But I think this is about the primary, do we risk having the primary on an older version than the snapshot (and yes we do and I agree that it is rare). So perhaps it should just read "allocation primaries" and not "allocating replicas" in this line?